test/fluree/db/query/fql_test.clj

```diff
@@ -319,32 +319,68 @@
             (is (= [["ex:bob"]] sut)
                 "returns correctly filtered results")))))))
 
-(deftest ^:integration ^:pending datatype-test
-  (let [conn   (test-utils/create-conn)
-        ledger @(fluree/create conn "people")
-        db     @(fluree/stage
-                  (fluree/db ledger)
-                  {"@context" ["https://ns.flur.ee"
-                               test-utils/default-context
-                               {:ex "http://example.org/ns/"}]
-                   "insert"
-                   [{:id      :ex/homer
-                     :ex/name "Homer"
-                     :ex/age  36}
-                    {:id      :ex/bart
-                     :ex/name "Bart"
-                     :ex/age  "forever 10"}]})]
-    (testing "including datatype in query results"
-      (let [query   {:context [test-utils/default-context
-                               {:ex "http://example.org/ns/"}]
-                     :select  '[?age ?dt]
-                     :where   '[{:ex/age ?age}
-                                [:bind ?dt (datatype ?age)]]}
-            results @(fluree/query db query)]
-        (is (= [["forever 10" "xsd:string"] [36 "xsd:long"]]
-               results))))
-    (testing "filtering query results with datatype fn")
-    (testing "filtering query results with @type value map")))
+(deftest ^:integration datatype-test
```

Comment:
I attempted to generate an iri and ran into an issue where we were expecting iris to be strings. Here's the setup:

```
(def conn @(fluree/connect {:method :memory}))
  (def ledger @(fluree/create conn "upserttest"))

  (def db0 (fluree/db ledger))

  (def db1 @(fluree/stage db0 {"insert"
                               [{"@id" "ex:dan"
                                 "@type" "ex:Person"
                                 "ex:email" "definitely@exists.com"}]}))

  (def db2 @(fluree/stage db1 {"where"
                               [["optional" {"@id" "?maybeFriend1" "ex:email" "definitely@exists.com"}]
                                ["optional" {"@id" "?maybeFriend2" "ex:email" "nobody@doesntexist.com"}]
                                ["bind" "?maybeFriend3" "(or ?maybeFriend2 (iri (concat \"ex:\" (struuid))))"]]
                               "insert"
                               [{"@id" "ex:hasFriend"
                                 "@type" "ex:Person"
                                 "ex:friend" "?maybeFriend1"}
                                {"@id" "ex:newFriend"
                                 "@type" "ex:Person"
                                 "ex:newProp" "yup"
                                 "ex:friend" "?maybeFriend3"}
                                {"@id" "?maybeFriend3"
                                 "ex:email" "nobody@doesntexist.com"}]}))
```

I can investigate further, but I thought I'd bookmark it here.

Reply:
There are going to be lots of places where we expect iris to be strings that we'll have to deal with in turn. That's what I was talking about in the description of this pr when I mentioned that we'd need to convert the existing strings into iris, and make existing code work with iris, gradually. This pr is just a first step that has the minimum functionality to make data type filtering work.

Reply:
I'm a bit hesitant to merge this in while things are actively broken, though - we typically keep `main` in a working state. Is it possible to have a list of things that need to be updated? That way we'll at least have scope outlined and we can keep track of the ways in which it's broken.

Reply:
Things are not "actively broken". the iri function is a new feature that explicitly only works right now in specific contexts. using the iri function anywhere outside of a where clause is not currently supported. we can add support later. Actively broken means something was working before and isn't working now. I have added a new feature, but it only works in specific contexts. We can expand that support later.

Reply:
This is in the where clause, though? I'm ok with merging it in as long as we know which contexts it does work in and which it doesn't, and I don't understand which is which.

Reply:
More precisely, you are using the _result_ of the iri function in the insert, and that's what's causing the problem.

Reply:
Ok, thanks for clearing that up. Can we make a ticket for the followup work so we can do it? I think Andrew's going to need it soon one way or the other : )